lib/src/parsers/parser.dart

```diff
@@ -0,0 +1,52 @@
+import 'package:vector_tile_renderer/src/expressions/argument_expression.dart';
+import 'package:vector_tile_renderer/src/expressions/expression.dart';
+import 'package:vector_tile_renderer/src/expressions/value_expression.dart';
+import 'package:vector_tile_renderer/src/logger.dart';
+import 'package:vector_tile_renderer/src/parsers/match.dart';
+
+import 'case.dart';
+import 'coalesce.dart';
+import 'step.dart';
+
+abstract class Parser<T> {
+  Expression<T>? parseSpecial(dynamic data);
+  Expression<T>? preCommon(dynamic data) {}
```

Comment:
The only use of preCommon seemed unused to I removed it entirely. I renamed the original `parse` function to `parseExpression` and `parseSpecial` to `parse`, then added `@protected` to the new `parse` function to have the analyzer warn callers that the function is only to be called by extending classes or internally and `@nonVirtual` from the `meta` package on `parseExpression` to warn if someone was to try and override that instead of `parse`.

So the caller is now warned when calling anything but `parseExpression` from the outside and the developer is warned when accidentally overriding the wrong function which should aid with this problem, I hope.